presto-main/src/main/java/com/facebook/presto/sql/planner/SortExpressionExtractor.java

```diff
@@ -82,71 +89,104 @@ private SortExpressionExtractor() {}
                     }
                     if (sortChannel.isPresent() && !hasBuildReferencesOnOtherSide) {
-                        return sortChannel.map(symbolReference -> new SortExpressionContext(symbolReference, singletonList(comparison)));
+                        return sortChannel.map(variableReference -> new SortExpressionContext(variableReference, singletonList(call)));
                     }
                     return Optional.empty();
                 default:
                     return Optional.empty();
             }
         }
+
+        @Override
+        public Optional<SortExpressionContext> visitInputReference(InputReferenceExpression input, Void context)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<SortExpressionContext> visitConstant(ConstantExpression literal, Void context)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<SortExpressionContext> visitLambda(LambdaDefinitionExpression lambda, Void context)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<SortExpressionContext> visitVariableReference(VariableReferenceExpression reference, Void context)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<SortExpressionContext> visitSpecialForm(SpecialFormExpression specialForm, Void context)
+        {
+            return Optional.empty();
+        }
     }
 
-    private static Optional<SymbolReference> asBuildSymbolReference(Set<Symbol> buildLayout, Expression expression)
+    private static Optional<VariableReferenceExpression> asBuildVariableReference(Set<Symbol> buildLayout, RowExpression expression)
     {
         // Currently only we support only symbol as sort expression on build side
-        if (expression instanceof SymbolReference) {
-            SymbolReference symbolReference = (SymbolReference) expression;
-            if (buildLayout.contains(new Symbol(symbolReference.getName()))) {
-                return Optional.of(symbolReference);
+        if (expression instanceof VariableReferenceExpression) {
+            VariableReferenceExpression reference = (VariableReferenceExpression) expression;
+            if (buildLayout.contains(new Symbol(reference.getName()))) {
+                return Optional.of(reference);
             }
         }
         return Optional.empty();
     }
 
-    private static boolean hasBuildSymbolReference(Set<Symbol> buildSymbols, Expression expression)
+    private static boolean hasBuildSymbolReference(Set<Symbol> buildSymbols, RowExpression expression)
     {
-        return new BuildSymbolReferenceFinder(buildSymbols).process(expression);
+        return expression.accept(new BuildSymbolReferenceFinder(buildSymbols), null);
     }
 
     private static class BuildSymbolReferenceFinder
```

Comment:
Could we replace this visitor with ```SymbolExtractor::extractUnique``` then check if result contains any symbol from buildSymbols?

Reply:
The space and time complexity of using `extractUnique` is higher than having a specialized visitor: `O(n)` vs `O(1)`